lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp

```diff
@@ -289,6 +289,48 @@ class DecomposeAtenExpandOp : public OpRewritePattern<AtenExpandOp> {
 };
 } // namespace
 
+// Decompose torch.mean into: sum(x)/div(numTensorElements).
+namespace {
+class DecomposeAtenMeanOp : public OpRewritePattern<AtenMeanOp> {
+public:
+  using OpRewritePattern::OpRewritePattern;
+  LogicalResult matchAndRewrite(AtenMeanOp op,
+                                PatternRewriter &rewriter) const override {
+    Location loc = op.getLoc();
+    Value input = op.self();
+    Value output = op.result();
+
+    Value dtype = rewriter.create<ConstantNoneOp>(loc);
+    Value sum = rewriter.create<AtenSumOp>(loc, output.getType(), input, dtype);
+    if (!sum)
+      return failure();
+
+    BaseTensorType inputTensorType = input.getType().cast<BaseTensorType>();
+    BaseTensorType outputTensorType = output.getType().cast<BaseTensorType>();
+    SmallVector<int64_t> sizes;
+    int64_t tensorSize = 1;
+    if (inputTensorType.hasSizes()) {
+      ArrayRef<int64_t> inputShape = inputTensorType.getSizes();
+      for (int64_t i : inputShape) {
+        if (i < 1)
+          return rewriter.notifyMatchFailure(op, "Invalid tensor size.");
+        tensorSize *= i;
+      }
+    } else
+      return rewriter.notifyMatchFailure(
+          op, "Only support input with known sizes.");
+
+    Value constTensorSize = rewriter.create<ConstantFloatOp>(
+        loc, rewriter.getF32FloatAttr((float)tensorSize));
+    Value divisorTensor = rewriter.create<PrimNumToTensorScalarOp>(
+        loc, outputTensorType, constTensorSize);
```

Comment:
note the result type of `PrimNumToTensorScalarOp` should always be <[],f64>.

Reply:
The result of `PrimNumToTensorScalarOp` is <[],f32> when the operand type is float. I have checked this in the RefineTypes.cpp

Reply:
I think the refine types of `NumToTensor` is wrong. Currently an e2e test of `NumToTensor` with a float input fails. The reason is that `torch.constant.float` gets turned into an `f64`, making the body of `NumToTensor` after `ConverTorchToLinalg` have the following form:

```mlir
/ -----// IR Dump After ConvertTorchToLinalg //----- //
func @forward(%arg0: !torch.vtensor<[1,3],f32>) -> !torch.vtensor<[],f32> {
  %float2.000000e00 = torch.constant.float 2.000000e+00
  %0 = torch_c.to_f64 %float2.000000e00
  %1 = linalg.init_tensor [] : tensor<f64>
  %2 = linalg.fill(%0, %1) : f64, tensor<f64> -> tensor<f64>
  %3 = torch_c.from_builtin_tensor %2 : tensor<f64> -> !torch.vtensor<[],f32>
  return %3 : !torch.vtensor<[],f32>
}
```

Notice that everything inside is an `f64`, but because of `RefineTypes`, the output of the function is set to `f32` tensor. This leads to an error downstream when the operation `%3 = torch_c.from_builtin_tensor %2 : tensor<f64> -> !torch.vtensor<[],f32>` is handled, because the element types don't match. (Note that a cast at the end of `CovertNumToTensor` does not fix this because you cannot cast to a different element type).

I believe this is the cause (or at least part of the cause) of the issue you're seeing @vivekkhandelwal1, for if you remove the `NumToTensor` op, for example

```diff
func @forward(%arg0: !torch.vtensor<[1,3],f32>) -> !torch.vtensor<[],f32> {
  %none = torch.constant.none
  %none_0 = torch.constant.none
  %0 = torch.aten.sum %arg0, %none_0 : !torch.vtensor<[1,3],f32>, !torch.none -> !torch.vtensor<[],f32>
  %float3.000000e00 = torch.constant.float 3.000000e+00
- %1 = torch.prim.NumToTensor.Scalar %float3.000000e00 : !torch.float -> !torch.vtensor<[],f32>
- %2 = torch.aten.div.Tensor %0, %1 : !torch.vtensor<[],f32>, !torch.vtensor<[],f32> -> !torch.vtensor<[],f32>
+ %2 = torch.aten.div.Tensor %0, %0 : !torch.vtensor<[],f32>, !torch.vtensor<[],f32> -> !torch.vtensor<[],f32>
  return %2 : !torch.vtensor<[],f32>
}
```

and run `./build/bin/torch-mlir-opt forward.mlir --pass-pipeline='torch-backend-to-linalg-on-tensors-backend-pipeline'` it works without the assert error.

Reply:
However, I also think that using `NumToTensor` is not the best option in this particular case because you don't want your `mean` to output an `f64` tensor when the input is an `f32` tensor. I think the more appropriate thing here would be to add a lowering for `AtenDivScalar` similar to `AtenMulScalar` 

https://github.com/llvm/torch-mlir/blob/1dc374014b4539651fdf36c83b896b93fd5703a5/lib/Conversion/TorchToLinalg/TorchToLinalg.cpp#L1577-L1587

and use that function instead, since that function will keep the tensor type consistent.

Reply:
Okay, so instead of `AtenDivTensorOp`, I should use `AtenDivScalarOp` (after adding its lowering). In that case, I would not need to create a `PrimNumToTensorScalarOp`. But still, I would need a `ConstantFloatOp` as a divisor. And `ConstantFloatOp` needs to have an F64 attribute, resulting in a type mismatch as the input to the mean and hence to the div will be F32.

Reply:
So the lowering for `AtenDivScalarOp` will take care of converting the `ConstantFloatOp` into an `f32`. This is done in the `AtenMulScalarOp` case in 

https://github.com/llvm/torch-mlir/blob/1dc374014b4539651fdf36c83b896b93fd5703a5/lib/Conversion/TorchToLinalg/TorchToLinalg.cpp#L1586

In other words, it converts the scalar argument to match the type of the tensor. This may seem strange, but it is the way PyTorch works. If a scalar is part of a tensor operation, such as `AtenDivScalarOp`, then the scalar float will get converted to an `f32` before the division happens, and then a promotion will be done so that the tensor and scalar match. If the tensor is `f32` nothing changes, if it is `f64` then the scalar will get promoted to `f64`. This has the net effect of always making the scalar convert to the tensor type (in the float case, at least).

On the other hand, if the scalar is not part of a Tensor op, such as `NumToTensor`, then the scalar follows a different set of rules. If it is a float, for example, it gets converted to `f64`.

You can see a discussion of this on the PyTorch side here

https://github.com/pytorch/pytorch/pull/67044#discussion_r734086719

Reply:
Thanks @ramiro050, I will do this once the PR for `AtenDivScalarOp` is merged.